relm4/src/sender.rs

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: MIT or Apache-2.0
+
+//! Contains various flavors of channels to send messages between components and workers.
+
+use crate::factory::FactoryComponent;
+use crate::{Component, Sender, ShutdownReceiver};
+use std::future::Future;
+
+/// Contains senders to send and receive messages from a [`Component`].
+#[derive(Debug)]
+pub struct ComponentSender<C: Component> {
+    /// Emits inputs to the component.
+    pub input: Sender<C::Input>,
+    /// Emits outputs from the component.
+    pub output: Sender<C::Output>,
+    command: Sender<C::CommandOutput>,
+    shutdown: ShutdownReceiver,
+}
+
+impl<C: Component> ComponentSender<C> {
+    pub(crate) fn new(
+        input_tx: Sender<C::Input>,
+        output_tx: Sender<C::Output>,
+        command_tx: Sender<C::CommandOutput>,
+        shutdown_tx: ShutdownReceiver,
+    ) -> Self {
+        ComponentSender {
+            input: input_tx,
+            output: output_tx,
+            command: command_tx,
+            shutdown: shutdown_tx,
+        }
+    }
+
+    /// Spawns a command.
+    /// You can bind the the command to the lifetime of the component
+    /// by using a [`ShutdownReceiver`].
+    pub fn command<Cmd, Fut>(&self, cmd: Cmd)
+    where
+        Cmd: FnOnce(Sender<C::CommandOutput>, ShutdownReceiver) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send,
+    {
+        let recipient = self.shutdown.clone();
+        let sender = self.command.clone();
+        crate::spawn(async move {
+            cmd(sender, recipient).await;
+        });
+    }
+
+    /// Spawns a future that will be dropped as soon as the component is shut down.
+    ///
+    /// Essentially, this is a simpler version of [`Self::command()`].
+    pub fn oneshot_command<Fut>(&self, future: Fut)
+    where
+        Fut: Future<Output = C::CommandOutput> + Send + 'static,
+    {
+        // Async closures would be awesome here...
+        self.command(move |out, shutdown| {
+            shutdown
+                .register(async move { out.send(future.await) })
+                .drop_on_shutdown()
+        });
+    }
+
+    /// Emit an input to the component.
+    pub fn input(&self, message: C::Input) {
+        self.input.send(message)
+    }
+
+    /// Equivalent to `&self.input`.
+    #[must_use]
+    pub fn input_sender(&self) -> &Sender<C::Input> {
+        &self.input
+    }
+
+    /// Emit an output to the component.
+    pub fn output(&self, message: C::Output) {
+        self.output.send(message)
+    }
+
+    /// Equivalent to `&self.output`.
+    #[must_use]
+    pub fn output_sender(&self) -> &Sender<C::Output> {
+        &self.output
+    }
+}
+
+impl<C: Component> Clone for ComponentSender<C> {
+    fn clone(&self) -> Self {
+        ComponentSender {
+            input: self.input.clone(),
+            output: self.output.clone(),
+            command: self.command.clone(),
+            shutdown: self.shutdown.clone(),
```

Comment:
I think this isn't great from a performance perspective. Most sender types use atomics internally so it would be better to have something similar to the old `Arc<ComponentSenderInner>` to avoid multiple atomic operations (it's not a big deal actually but should be an easy optimization).

Reply:
Agreed, fixed by adding back `ComponentSenderInner`.